File: web/js/vision/faceLandmarker.js
// MediaPipe FaceLandmarker (Web) wrapper — components 2a + 2b.
//
// Browser equivalent of `vision/face_landmarker.py`. Runs the pre-trained
// face_landmarker.task model on a <video> frame, selects the LARGEST face,
// applies the confidence/size gate, and returns a plain FaceDetection-like
// object ({bbox, landmarks, confidence}) or null when the frame is ignored.
//
// The .task model file is the SAME bundle Google publishes for the Python and
// Android APIs, so the landmark indices used by lips.js are identical.
//
// This is the ONE place that imports MediaPipe. Swap this file (keeping the
// detect(video, tsMs) contract) to use a different foundation model.

// Default CDN pins. Overridable via config for offline/self-hosting.
const TASKS_VISION_VERSION = "0.10.21"; // matches Python mediapipe==0.10.21
export const DEFAULT_WASM_BASE = `https://cdn.jsdelivr.net/npm/@mediapipe/tasks-vision@${TASKS_VISION_VERSION}/wasm`;
export const DEFAULT_MODEL_URL =
  "https://storage.googleapis.com/mediapipe-models/face_landmarker/" +
  "face_landmarker/float16/1/face_landmarker.task";

export class FaceLandmarkerEngine {
  /**
   * @param {Object} [opts]
   * @param {number} [opts.minFaceConfidence=0.5]
   * @param {number} [opts.minFaceAreaRatio=0.02]
   * @param {number} [opts.numFaces=3]
   * @param {string} [opts.modelUrl]
   * @param {string} [opts.wasmBase]
   */
  constructor(opts = {}) {
    this.minFaceConfidence = opts.minFaceConfidence ?? 0.5;
    this.minFaceAreaRatio = opts.minFaceAreaRatio ?? 0.02;
    this.numFaces = opts.numFaces ?? 3;
    this.modelUrl = opts.modelUrl ?? DEFAULT_MODEL_URL;
    this.wasmBase = opts.wasmBase ?? DEFAULT_WASM_BASE;
    this._landmarker = null;
  }

  /** Load the WASM fileset + model. ~3.7 MB model, fetched once (SW-cacheable). */
  async init() {
    const { FaceLandmarker, FilesetResolver } = await import(
      /* @vite-ignore */
      `https://cdn.jsdelivr.net/npm/@mediapipe/tasks-vision@${TASKS_VISION_VERSION}`
    );
    const fileset = await FilesetResolver.forVisionTasks(this.wasmBase);
    const opts = (delegate) => ({
      baseOptions: { modelAssetPath: this.modelUrl, delegate },
      runningMode: "VIDEO",
      numFaces: this.numFaces,
      minFaceDetectionConfidence: this.minFaceConfidence,
      minFacePresenceConfidence: this.minFaceConfidence,
      minTrackingConfidence: this.minFaceConfidence,
      outputFaceBlendshapes: false,
      outputFacialTransformationMatrixes: false,
    });
    // Prefer GPU; fall back to CPU on devices/browsers that reject the delegate.
    try {
      this._landmarker = await FaceLandmarker.createFromOptions(fileset, opts("GPU"));
    } catch (err) {
      console.warn("GPU delegate unavailable, falling back to CPU:", err);
      this._landmarker = await FaceLandmarker.createFromOptions(fileset, opts("CPU"));
    }
  }

  /**
   * Detect the gated largest face for a video frame.
   * @param {HTMLVideoElement} video
   * @param {number} timestampMs monotonic ms (must strictly increase)
   * @returns {?{bbox:Object, landmarks:number[][], confidence:number}}
   */
  detect(video, timestampMs) {
    if (!this._landmarker) throw new Error("FaceLandmarkerEngine.init() not called");
    const result = this._landmarker.detectForVideo(video, timestampMs);
    const faces = result?.faceLandmarks;
    if (!faces || faces.length === 0) return null; // none passed the model gate (2a)

    let bestPoints = null;
    let bestBbox = null;
    let bestArea = -1.0;
    for (const landmarks of faces) {
      const points = landmarks.map((lm) => [lm.x, lm.y, lm.z]);
      const bbox = bboxFromPoints(points);
      if (bbox.area > bestArea) {
        bestArea = bbox.area;
        bestPoints = points;
        bestBbox = bbox;
      }
    }

    // Size gate: ignore faces whose box is too small a fraction of the frame (2a).
    if (bestBbox.area < this.minFaceAreaRatio) return null;

    return { bbox: bestBbox, landmarks: bestPoints, confidence: 1.0 };
  }

  close() {
    try {
      this._landmarker?.close();
    } catch {
      /* ignore */
    }
    this._landmarker = null;
  }
}

function bboxFromPoints(points) {
  let x0 = Infinity, y0 = Infinity, x1 = -Infinity, y1 = -Infinity;
  for (const [x, y] of points) {
    if (x < x0) x0 = x;
    if (x > x1) x1 = x;
    if (y < y0) y0 = y;
    if (y > y1) y1 = y;
  }
  const width = x1 - x0;
  const height = y1 - y0;
  return { x: x0, y: y0, width, height, area: Math.max(0, width) * Math.max(0, height) };
}

File: breathing_monitor/vision/face_landmarker.py
"""MediaPipe FaceLandmarker wrapper — components 2a + 2b.

Runs the pre-trained ``face_landmarker.task`` model on a single RGB frame,
selects the **largest** face, applies the confidence/size gate, and returns a
plain :class:`FaceDetection` (or ``None`` when the frame should be ignored).

Confidence gating note: the MediaPipe Tasks FaceLandmarker does not surface a
per-face score in its Python result; instead it drops faces below
``min_face_detection_confidence`` / ``min_face_presence_confidence`` internally.
We pass the user's threshold through to those options (so "ignore low-confidence
faces" is honored at the model level) and additionally gate on bounding-box
area here. The reported ``confidence`` field is therefore a pass/near-1.0
sentinel — the meaningful gate is the threshold + the area check.
"""

from __future__ import annotations

from pathlib import Path
from typing import List, Optional

import numpy as np

from ..core.types import BoundingBox, FaceDetection, Point
from .model import ensure_model


class FaceLandmarkerEngine:
    """Thin, reusable wrapper around MediaPipe's FaceLandmarker (IMAGE mode)."""

    def __init__(
        self,
        min_face_confidence: float = 0.5,
        min_face_area_ratio: float = 0.02,
        model_path: Optional[Path] = None,
        allow_download: bool = True,
    ) -> None:
        # Imported lazily so the pure-core tests never need mediapipe installed.
        import mediapipe as mp
        from mediapipe.tasks import python as mp_python
        from mediapipe.tasks.python import vision as mp_vision

        self._mp = mp
        self.min_face_area_ratio = min_face_area_ratio

        path = model_path or ensure_model(allow_download=allow_download)
        base_options = mp_python.BaseOptions(model_asset_path=str(path))
        options = mp_vision.FaceLandmarkerOptions(
            base_options=base_options,
            running_mode=mp_vision.RunningMode.IMAGE,
            num_faces=3,  # detect a few, then we pick the largest
            min_face_detection_confidence=min_face_confidence,
            min_face_presence_confidence=min_face_confidence,
            output_face_blendshapes=False,
            output_facial_transformation_matrixes=False,
        )
        self._landmarker = mp_vision.FaceLandmarker.create_from_options(options)

    def detect(self, rgb_image: np.ndarray) -> Optional[FaceDetection]:
        """Return the gated largest face for an RGB uint8 image, or ``None``."""
        mp_image = self._mp.Image(
            image_format=self._mp.ImageFormat.SRGB, data=np.ascontiguousarray(rgb_image)
        )
        result = self._landmarker.detect(mp_image)
        faces = getattr(result, "face_landmarks", None)
        if not faces:
            return None  # no face passed the model's confidence gate (2a)

        best_points: Optional[List[Point]] = None
        best_bbox: Optional[BoundingBox] = None
        best_area = -1.0
        for landmarks in faces:
            points: List[Point] = [(lm.x, lm.y, lm.z) for lm in landmarks]
            bbox = _bbox_from_points(points)
            if bbox.area > best_area:
                best_area = bbox.area
                best_points = points
                best_bbox = bbox

        assert best_points is not None and best_bbox is not None
        # Size gate: ignore faces whose box is too small a fraction of frame (2a).
        if best_bbox.area < self.min_face_area_ratio:
            return None

        return FaceDetection(bbox=best_bbox, landmarks=best_points, confidence=1.0)

    def close(self) -> None:
        try:
            self._landmarker.close()
        except Exception:
            pass


def _bbox_from_points(points: List[Point]) -> BoundingBox:
    xs = [p[0] for p in points]
    ys = [p[1] for p in points]
    x0, x1 = min(xs), max(xs)
    y0, y1 = min(ys), max(ys)
    return BoundingBox(x=x0, y=y0, width=x1 - x0, height=y1 - y0)
